### PR TITLE
Dynamic block sigops and max block size

### PR DESCRIPTION
**Pull request by Naviabheeman**
Block size and block sig ops limits are made dynamic depending on max block size which is read from xfield history

### CODE REVIEW

File: src/consensus/consensus.h [hunk 1]
```diff
@@ -14,9 +14,8 @@
   * The maximum allowed size for a block, in bytes (network rule).
   */
 static const unsigned int MAX_BLOCK_SIZE = 1000000;
-/** 
- * The maximum allowed number of signature check operations in a block (network rule) */
-static const int64_t MAX_BLOCK_SIGOPS = MAX_BLOCK_SIZE/50;
+/** Default blockmax size, which controls the range of block size the mining code will create **/
+static const unsigned int DEFAULT_BLOCK_MAX_SIZE = MAX_BLOCK_SIZE - 1000;
```
Comment (azuchi) [inline src/consensus/consensus.h:18]: What is the difference between `MAX_BLOCK_SIZE` and `DEFAULT_BLOCK_MAX_SIZE`?

File: src/consensus/tx_verify.cpp [hunk 2]
```diff
@@ -135,9 +136,9 @@ unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& in
     return nSigOps;
 }
 
-int64_t GetTransactionSigOps(const CTransaction& tx, const CCoinsViewCache& inputs, int flags)
+int32_t GetTransactionSigOps(const CTransaction& tx, const CCoinsViewCache& inputs, int flags)
```
Comment (azuchi) [inline src/consensus/tx_verify.cpp:139]: What are the reasons for this type of change?

File: src/xfieldhistory.cpp [hunk 3]
```diff
@@ -82,4 +82,14 @@ bool IsXFieldNew(const CXField& xfield, CXFieldHistoryMap* pxfieldHistory)
     return xfield.IsValid()
           && std::find(XFIELDTYPES_INIT_LIST.begin(), XFIELDTYPES_INIT_LIST.end(), xfield.xfieldType) != XFIELDTYPES_INIT_LIST.end()
           && !boost::apply_visitor(checkVisitor, xfield.xfieldValue);
-}
\ No newline at end of file
+}
+
+uint32_t CURRENT_MAXBLOCKSIZE() {
```
Comment (azuchi) [inline src/xfieldhistory.cpp:87]: These upper snake cases are easily mistaken for constants, so the function name should be camel case as well as others.

File: test/functional/feature_xfield_maxblocksize.py [hunk 4]
```diff
@@ -32,7 +32,7 @@
 from test_framework.script import CScript, OP_CHECKSIG, OP_TRUE, SignatureHash, SIGHASH_ALL
 from test_framework.messages import CTransaction, MAX_BLOCK_BASE_SIZE, CTxOut, CTxIn, COutPoint, uint256_from_str, ser_compact_size, msg_headers, CBlockHeader
 
-MAX_BLOCK_SIGOPS = 20000
+MAX_BLOCK_SIGOPS = int((100000 - 1000) / 50)
```
Comment (azuchi) [inline test/functional/feature_xfield_maxblocksize.py:35]: What is the basis for this formula?

**Review by azuchi — CHANGES_REQUESTED** [review]

File: test/functional/feature_xfield_maxblocksize.py (hunk 4 above)
Comment (Naviabheeman) [reply test/functional/feature_xfield_maxblocksize.py:35]: 20000 would be the value of  MAX_BLOCK_SIZE/50. we now need it to be DEFAULT_BLOCK_MAX_SIZE/50 as DEFAULT_BLOCK_MAX_SIZE is the initial value in XFieldHistory for MaxBlockSize

File: src/consensus/tx_verify.cpp (hunk 2 above)
Comment (Naviabheeman) [reply src/consensus/tx_verify.cpp:139]: Our max block size xfield is only 4 bytes.  Transaction Sigops, block signups are all derived from it. So I made them all 4 bytes.

**Review by azuchi — CHANGES_REQUESTED** [review]

**Comment by azuchi** [comment]
@Naviabheeman In the following code, MAX_BLOCK_SIZE is referenced, is this a problem?

in `coins.cpp`:

```cpp
static const size_t MAX_OUTPUTS_PER_BLOCK = MAX_BLOCK_SIZE / MIN_TRANSACTION_OUTPUT_SIZE;

const Coin& AccessByTxid(const CCoinsViewCache& view, const uint256& txid)
{
    COutPoint iter(txid, 0);
    while (iter.n < MAX_OUTPUTS_PER_BLOCK) {
        const Coin& alternate = view.AccessCoin(iter);
        if (!alternate.IsSpent()) return alternate;
        ++iter.n;
    }
    return coinEmpty;
}
```

in `tapyrus-tx.cpp`:

```cpp
static const unsigned int maxVout = MAX_BLOCK_SIZE / minTxOutSz;
```

**Comment by Naviabheeman** [comment]
> ```c++
> static const size_t MAX_OUTPUTS_PER_BLOCK = MAX_BLOCK_SIZE / MIN_TRANSACTION_OUTPUT_SIZE;
> ```

These were intentionally left as they were. `MAX_OUTPUTS_PER_BLOCK` is guessing the maximum possible vout and using it to get any unspent output from coins db. If we make this dynamic the range will be huge and make the worse case scenario i.e. when transaction does not have unspent outputs even worse.
On the other hand if we leave this as it is, only two places are affected:   `GetTransaction` RPC and retrieving transaction undo data during a reorg. In case a the block size becomes bigger, say 3MB and there is a huge transaction with more than MAX_OUTPUTS_PER_BLOCK outputs, then gettransaction RPC on that transaction will fail. And in case this huge transaction is reversed in a reorg then the undo input may not be available. 
Since these two scenarios are rare I think we can leave MAX_OUTPUTS_PER_BLOCK limit as it is. Adding a new API to level db wrapper which can get all the outputs of a transaction would be a better solution.

**Pull request merged**